File: src-model/model/Coords.java
package model;

import engine.utils.TextLogger;

/**
 * Represents mutable coordinates in 3D space.
 */
public final class Coords {

    private double x;
    private double y;
    private double z;

    /* ========== PUBLIC ========== */
    /** Used to create coordinates. */
    public Coords(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    /** Used to clone coordinates. */
    public Coords(Coords coords) {
        this.x = coords.x;
        this.y = coords.y;
        this.z = coords.z;
    }

    @Override
    public String toString() {
        return "(" + TextLogger.twoDecimal(x) + ", " + TextLogger.twoDecimal(y) + ", " + TextLogger.twoDecimal(z) + ")";
    }

    /* ========== PRIVATE ========== */
    /* ========== PROPERTIES ========== */
    public double getX() {
        return x;
    }

    public void setX(double x) {
        this.x = x;
    }

    public double getY() {
        return y;
    }

    public void setY(double y) {
        this.y = y;
    }

    public double getZ() {
        return z;
    }

    public void setZ(double z) {
        this.z = z;
    }
}

File: src-model/model/Coords.kt
package model

import engine.utils.TextLogger

class Coords(
        var x: Float = 0F,
        var y: Float = 0F,
        var z: Float = 0F
) {

    // TODO consider making it immutable
    operator fun plusAssign(coords: Coords) {
        x += coords.x
        y += coords.y
        z += coords.z
    }

    fun distanceTo(coords: Coords): Float {
        val x = Math.pow((x - coords.x).toDouble(), 2.0)
        val y = Math.pow((y - coords.y).toDouble(), 2.0)
        val z = Math.pow((z - coords.z).toDouble(), 2.0)

        return Math.sqrt(x + y + z).toFloat()
    }

    override fun toString(): String {
        return "(${TextLogger.twoDecimal(x)}, ${TextLogger.twoDecimal(y)}, ${TextLogger.twoDecimal(z)})"
    }
}